### PR TITLE
Add payload validation for API v2

### LINKED ISSUE #3491: v2 API is missing payload validations when creating a flow
In v1 there's a number of validations applied to the flow payload, for example: flowId cannot be empty or port number/vlan cannot be negative etc..
In v2 all these validations are missing, allowing to create broken flows. Existence of such flows in the database also breaks flow-related v1 APIs.

### PR DESCRIPTION
**Pull request by sergii-iakovenko**
Resolves #3491

### CODE REVIEW

**Review by dmitrii-beliakov — APPROVED** [review]

**Review by niksv — APPROVED** [review]

**Comment by pkazlenka** [comment]
@durgakako , I've found, that validation for YFlow create operation allows to request YFlow with negative shared endpoint port number. According to Kibana logs, such a request successfully pass all the validations, then kilda tries to install rules to switch with negative port number (-1 in the test example) and silently fails, not creating any flow after switch rejects to accept incorrect rule.
Please, fix.

**Comment by pkazlenka** [comment]
Another problem: new errors in validation aren't handled as the existing ones:
1. Create valid Y-FLow
2. Send update Y-Flow request with `shared_endpoint = null`
=> Error message received: 'JSON parse error: Cannot construct instance of `org.openkilda.northbound.dto.v2.yflows.YFlowUpdatePayload`, problem: sharedEndpoint is marked non-null but is null;`
Expected: Error message is `Could not update y-flow` with description message like `Shared Endpoint can't be null`

**Comment by durgakako** [comment]
> @durgakako , I've found, that validation for YFlow create operation allows to request YFlow with negative shared endpoint port number. According to Kibana logs, such a request successfully pass all the validations, then kilda tries to install rules to switch with negative port number (-1 in the test example) and silently fails, not creating any flow after switch rejects to accept incorrect rule.
> Please, fix.
> I can't attach .json here, so I'll pass it to you directly.

Fixed it

**Comment by durgakako** [comment]
> @durgakako , I've found, that validation for YFlow create operation allows to request YFlow with negative shared endpoint port number. According to Kibana logs, such a request successfully pass all the validations, then kilda tries to install rules to switch with negative port number (-1 in the test example) and silently fails, not creating any flow after switch rejects to accept incorrect rule.
> Please, fix.
> I can't attach .json here, so I'll pass it to you directly.

fixed.

**Comment by pkazlenka** [comment]
@durgakako 
Error for flow create/update validation is not adhere to kilda standards now. E.g., when I try to update regular flow with `source.vlanId = 10000`, kilda returns error 
```
{"errors":{"source.vlanId":"vlanId can't be greater than 4095"}}
```
while all the HTTP responses with validation errors must have payload following the model from `src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/MessageError.java`, e.g. :
```
{"correlation_id":"01f77a7f-7699-4ebf-b240-17fdb37de2aa : fn-tests-f2d7cfc9-3bd9-4670-8ec6-a72308ececc1",
"timestamp":1681734962207,
"error-type":"Object was not found",
"error-message":"Could not update flow",
"error-description":"Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:01 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=9223372036854775807"}
```

**Comment by durgakako** [comment]
> Another problem: new errors in validation aren't handled as the existing ones:
> 
> 1. Create valid Y-FLow
> 2. Send update Y-Flow request with `shared_endpoint = null`
>    => Error message received: 'JSON parse error: Cannot construct instance of `org.openkilda.northbound.dto.v2.yflows.YFlowUpdatePayload`, problem: sharedEndpoint is marked non-null but is null;`Expected: Error message is`Could not update y-flow`with description message like`Shared Endpoint can't be null`

Updated.

**Comment by durgakako** [comment]
> @durgakako Error for flow create/update validation is not adhere to kilda standards now. E.g., when I try to update regular flow with `source.vlanId = 10000`, kilda returns error
> 
> ```
> {"errors":{"source.vlanId":"vlanId can't be greater than 4095"}}
> ```
> 
> while all the HTTP responses with validation errors must have payload following the model from `src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/error/MessageError.java`, e.g. :
> 
> ```
> {"correlation_id":"01f77a7f-7699-4ebf-b240-17fdb37de2aa : fn-tests-f2d7cfc9-3bd9-4670-8ec6-a72308ececc1",
> "timestamp":1681734962207,
> "error-type":"Object was not found",
> "error-message":"Could not update flow",
> "error-description":"Not enough bandwidth or no path found. Switch 00:00:00:00:00:00:00:01 doesn't have links with enough bandwidth, Failed to find path with requested bandwidth=9223372036854775807"}
> ```
> 
> Same for y-flow create/update errors.

Updated.

**Comment by durgakako** [comment]
> @durgakako , we need one more round, I'm afraid.
> 
> 1. When you can't create a Y-Flow for any reason, the error message should say `Could not create y-flow` and the error description must explain in details, e.g., as you do now, `Errors:PortNumber must be non-negative`.
> 2. Please, don't use new lines in error message/description. Split reasons with comma.
> 3. For update/partial update, error message should say `Could not update y-flow`.

updated:

{
  "correlation_id": "59df5616-2492-4c84-8d0d-b13afbca87f8 : Durga_1",
  "timestamp": 1682536289579,
  "error-type": "Invalid request data",
  "error-message": "Could not create y-flow",
  "error-description": "Errors: VlanId must be non-negative, PortNumber must be non-negative"
}

**Comment by pkazlenka** [comment]
@durgakako , tried the latest version. While error responses look awesome, there's some problem with behavior:
I receive HTTP/500 when try to perform y-flow update, making `shared_endpoint = null`. I can't trace the cause of the problem, because search in kibana by `correlation_id` returns only request and response, without any intermediate calls :
image

Here's update payload:
```
{
  "shared_endpoint" : null,
  "maximum_bandwidth" : 1000,
  "path_computation_strategy" : "cost",
  "encapsulation_type" : "transit_vlan",
  "max_latency" : null,
  "max_latency_tier2" : null,
  "ignore_bandwidth" : false,
  "periodic_pings" : false,
  "pinned" : false,
  "priority" : null,
  "strict_bandwidth" : false,
  "description" : "autotest y-flow: How bitter a thing it is to look into happiness through another man's eyes!",
  "allocate_protected_path" : false,
  "diverse_flow_id" : null,
  "sub_flows" : [ {
    "flow_id" : "ys-kyiq4gpe44i63oxwigrfjtvgfy",
    "endpoint" : {
      "switch_id" : "00:00:00:00:00:00:00:01",
      "port_number" : 9,
      "vlan_id" : 3198,
      "inner_vlan_id" : 0
    },
    "shared_endpoint" : {
      "vlan_id" : 726,
      "inner_vlan_id" : 0
    },
    "status" : "Up",
    "description" : null,
    "time_create" : "2023-04-27T10:36:15.207Z",
    "time_update" : "2023-04-27T10:36:16.345Z"
  }, {
    "flow_id" : "ys-kyiq4gxe44i63oxwigrfjtvgfy",
    "endpoint" : {
      "switch_id" : "00:00:00:00:00:00:00:01",
      "port_number" : 17,
      "vlan_id" : 1152,
      "inner_vlan_id" : 0
    },
    "shared_endpoint" : {
      "vlan_id" : 1755,
      "inner_vlan_id" : 0
    },
    "status" : "Up",
    "description" : null,
    "time_create" : "2023-04-27T10:36:15.464Z",
    "time_update" : "2023-04-27T10:36:16.594Z"
  } ]
}
```

**Comment by durgakako** [comment]
> @durgakako , tried the latest version. While error responses look awesome, there's some problem with behavior: I receive HTTP/500 when try to perform y-flow update, making `shared_endpoint = null`. I can't trace the cause of the problem, because search in kibana by `correlation_id` returns only request and response, without any intermediate calls : image
> 
> Here's update payload:
> 
> ```
> {
>   "shared_endpoint" : null,
>   "maximum_bandwidth" : 1000,
>   "path_computation_strategy" : "cost",
>   "encapsulation_type" : "transit_vlan",
>   "max_latency" : null,
>   "max_latency_tier2" : null,
>   "ignore_bandwidth" : false,
>   "periodic_pings" : false,
>   "pinned" : false,
>   "priority" : null,
>   "strict_bandwidth" : false,
>   "description" : "autotest y-flow: How bitter a thing it is to look into happiness through another man's eyes!",
>   "allocate_protected_path" : false,
>   "diverse_flow_id" : null,
>   "sub_flows" : [ {
>     "flow_id" : "ys-kyiq4gpe44i63oxwigrfjtvgfy",
>     "endpoint" : {
>       "switch_id" : "00:00:00:00:00:00:00:01",
>       "port_number" : 9,
>       "vlan_id" : 3198,
>       "inner_vlan_id" : 0
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 726,
>       "inner_vlan_id" : 0
>     },
>     "status" : "Up",
>     "description" : null,
>     "time_create" : "2023-04-27T10:36:15.207Z",
>     "time_update" : "2023-04-27T10:36:16.345Z"
>   }, {
>     "flow_id" : "ys-kyiq4gxe44i63oxwigrfjtvgfy",
>     "endpoint" : {
>       "switch_id" : "00:00:00:00:00:00:00:01",
>       "port_number" : 17,
>       "vlan_id" : 1152,
>       "inner_vlan_id" : 0
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 1755,
>       "inner_vlan_id" : 0
>     },
>     "status" : "Up",
>     "description" : null,
>     "time_create" : "2023-04-27T10:36:15.464Z",
>     "time_update" : "2023-04-27T10:36:16.594Z"
>   } ]
> }
> ```

Fixed.

**Comment by pkazlenka** [comment]
> Fixed.

Update works great now, but create is broken again. Instead of receiving error, I receive HTTP/201, but then flow silently fails to be created (again, without any mentions in kibana, at least with correlation_id). The create request:
```
{
  "y_flow_id" : null,
  "shared_endpoint" : {
    "switch_id" : "00:00:00:22:3d:6c:00:40",
    "port_number" : -1
  },
  "maximum_bandwidth" : 1000,
  "path_computation_strategy" : "COST",
  "encapsulation_type" : "TRANSIT_VLAN",
  "max_latency" : null,
  "max_latency_tier2" : null,
  "ignore_bandwidth" : false,
  "periodic_pings" : false,
  "pinned" : false,
  "priority" : null,
  "strict_bandwidth" : false,
  "description" : "autotest y-flow: A horse! a horse! my kingdom for a horse!.",
  "allocate_protected_path" : false,
  "diverse_flow_id" : null,
  "sub_flows" : [ {
    "flow_id" : null,
    "endpoint" : {
      "switch_id" : "00:00:00:22:3d:5a:04:87",
      "port_number" : 15,
      "vlan_id" : 3550,
      "inner_vlan_id" : 0
    },
    "shared_endpoint" : {
      "vlan_id" : 1485,
      "inner_vlan_id" : 0
    },
    "description" : null
  }, {
    "flow_id" : null,
    "endpoint" : {
      "switch_id" : "00:00:00:22:3d:5a:04:87",
      "port_number" : 20,
      "vlan_id" : 856,
      "inner_vlan_id" : 0
    },
    "shared_endpoint" : {
      "vlan_id" : 2938,
      "inner_vlan_id" : 0
    },
    "description" : null
  } ]
}
```

**Comment by durgakako** [comment]
> > Fixed.
> 
> Update works great now, but create is broken again. Instead of receiving error, I receive HTTP/201, but then flow silently fails to be created (again, without any mentions in kibana, at least with correlation_id). The create request:
> 
> ```
> {
>   "y_flow_id" : null,
>   "shared_endpoint" : {
>     "switch_id" : "00:00:00:22:3d:6c:00:40",
>     "port_number" : -1
>   },
>   "maximum_bandwidth" : 1000,
>   "path_computation_strategy" : "COST",
>   "encapsulation_type" : "TRANSIT_VLAN",
>   "max_latency" : null,
>   "max_latency_tier2" : null,
>   "ignore_bandwidth" : false,
>   "periodic_pings" : false,
>   "pinned" : false,
>   "priority" : null,
>   "strict_bandwidth" : false,
>   "description" : "autotest y-flow: A horse! a horse! my kingdom for a horse!.",
>   "allocate_protected_path" : false,
>   "diverse_flow_id" : null,
>   "sub_flows" : [ {
>     "flow_id" : null,
>     "endpoint" : {
>       "switch_id" : "00:00:00:22:3d:5a:04:87",
>       "port_number" : 15,
>       "vlan_id" : 3550,
>       "inner_vlan_id" : 0
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 1485,
>       "inner_vlan_id" : 0
>     },
>     "description" : null
>   }, {
>     "flow_id" : null,
>     "endpoint" : {
>       "switch_id" : "00:00:00:22:3d:5a:04:87",
>       "port_number" : 20,
>       "vlan_id" : 856,
>       "inner_vlan_id" : 0
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 2938,
>       "inner_vlan_id" : 0
>     },
>     "description" : null
>   } ]
> }
> ```

Please could check on correct branch, I have tested got correct result for above payload: 

{
  "correlation_id": "a98b0ccd-a0ef-4350-8b5f-3423e13942cc : Durga-1",
  "timestamp": 1682691993084,
  "error-type": "Invalid request data",
  "error-message": "Could not create y-flow",
  "error-description": "Errors: PortNumber must be non-negative"
}

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 1]
```diff
@@ -368,7 +368,6 @@ class QinQFlowSpec extends HealthCheckSpecification {
         then: "Human readable error is returned"
         def exc = thrown(HttpClientErrorException)
         exc.statusCode == HttpStatus.BAD_REQUEST
-        exc.responseBodyAsString.to(MessageError).errorMessage == "Invalid request payload"
```
Comment (pkazlenka) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:371]: Don't do like this in the future, please.

**Pull request merged**